src/rt-app_parse_config.c

```diff
@@ -822,6 +823,87 @@ static sched_data_t *parse_sched_data(struct json_object *obj, int def_policy)
 	return NULL;
 }
 
+static size_t parse_taskgroup_name(char *src, char *dst, int copy)
+{
+	size_t len = 1;
+
+	/* Enforce leading slash. */
+	*dst = '/';
+
+	for (; *src != '\0'; src++)
+		/* Reduce sequence of slashes to one slash. */
+		if (*src != '/' || *src != *dst) {
+			if (copy)
+				dst++;
+			/*
+			 * Set *dst to *src even in 'copy == 0' case for
+			 * 'sequence of slashes' detection.
+			 */
+			*dst = *src;
+			len++;
+		}
+
+	/* Remove trailing slash except for root taskgroup. */
+	if (len == 1 || *dst != '/') {
+		if (copy)
+			dst++;
+		len++;
+	}
+
+	*dst = '\0';
+
+	return len;
+}
+
+static taskgroup_data_t *parse_taskgroup_data(struct json_object *obj)
+{
+	char *name = get_string_value_from(obj, "taskgroup", TRUE, "");
+	taskgroup_data_t *tg;
+	size_t size;
+	char dummy;
+
+	if (!strlen(name))
+		return NULL;
+
+	tg = find_taskgroup(name);
```

Comment:
find_taskgroup uses strcmp() to check if the group already exist but the name that is saved in ctrl.tg_array has been potentially modified. If "taskgroup" : "/tg1//tg11" is used in a json file, "/tg1/tg11" (with only 1 /) will be saved so if "taskgroup" : "/tg1//tg11" is used in several place in the json file, a new element will be used in ctrl.tg_array for each  "taskgroup" : "/tg1//tg11".
I have been able to reach the 32 limits with using always "taskgroup" : "/tg1//tg11" but in several different places of the json file whereas we should have used the same element of ctrl.tg_array

Reply:
Yes, you're right! There is this problem. Let me test if not removing the double slashes and the one at the end as well as not forcing a start slash causes any issue. If this is the case I could also abort the config parsing and leave the issue to the user ti fix it.
I will definitely add your testcase to my test suite. Thanks for the review and the testing!

Reply:
If you have issue when not removing multiple slashes and the one at the end or not forcing a start slash, you can probably just raise an error and let the user fixes the json file as you proposed instead of trying to fix it

Reply:
I agree. Can't fix it today since I'm traveling but will look into it on Monday.

Reply:
So this issue can be fixed by (1) enforcing a leading slash in the json file and (2) by adapting cgroup_mkdir() and cgroup_rmdir to deal with multiple adjacent slashes.

(1) goes into patch 'taskgroups: Config parser'
```
index aa82b022e504..7a54024750bd 100644
--- a/src/rt-app_parse_config.c
+++ b/src/rt-app_parse_config.c
@@ -823,61 +823,29 @@ static sched_data_t *parse_sched_data(struct json_object *obj, int def_policy)
        return NULL;
 }
 
-static size_t parse_taskgroup_name(char *src, char *dst, int copy)
-{
-       size_t len = 1;
-
-       /* Enforce leading slash. */
-       *dst = '/';
-
-       for (; *src != '\0'; src++)
-               /* Reduce sequence of slashes to one slash. */
-               if (*src != '/' || *src != *dst) {
-                       if (copy)
-                               dst++;
-                       /*
-                        * Set *dst to *src even in 'copy == 0' case for
-                        * 'sequence of slashes' detection.
-                        */
-                       *dst = *src;
-                       len++;
-               }
-
-       /* Remove trailing slash except for root taskgroup. */
-       if (len == 1 || *dst != '/') {
-               if (copy)
-                       dst++;
-               len++;
-       }
-
-       *dst = '\0';
-
-       return len;
-}
-
 static taskgroup_data_t *parse_taskgroup_data(struct json_object *obj)
 {
        char *name = get_string_value_from(obj, "taskgroup", TRUE, "");
        taskgroup_data_t *tg;
-       size_t size;
-       char dummy;
 
        if (!strlen(name))
                return NULL;
 
+       if (*name != '/') {
+               log_critical(PIN2 "Taskgroup [%s] has to start with [/]", name);
+               exit(EXIT_INV_CONFIG);
+       }
+
        tg = find_taskgroup(name);
        if (tg)
                return tg;
-
-       size = parse_taskgroup_name(name, &dummy, 0);
-
-       tg = alloc_taskgroup(size);
+       tg = alloc_taskgroup(strlen(name) + 1);
        if (!tg) {
                log_critical(PIN2 "Cannot allocate taskgroup");
                exit(EXIT_FAILURE);
        }
 
-       parse_taskgroup_name(name, tg->name, 1);
+       strcpy(tg->name, name);
 
        return tg;
 }
```
(2) goes into patch 'taskgroups: Cgroups functions'

```diff --git a/src/rt-app_taskgroups.c b/src/rt-app_taskgroups.c
index 4c92c4ff57dd..4fe3ae2d0202 100644
--- a/src/rt-app_taskgroups.c
+++ b/src/rt-app_taskgroups.c
@@ -228,6 +228,9 @@ static int cgroup_mkdir(const char *name, int *offset)
                while (dir[i] == '/')
                        i++;
 
+               if (dir[i] == '\0')
+                       goto error;
+
                while (dir[i] != '\0' && dir[i] != '/')
                        i++;
 
@@ -263,7 +266,7 @@ static int cgroup_rmdir(char *name, int offset)
 {
        char *dir = NULL, *path = NULL, *sep;
        int ret = 0;
-       size_t size;
+       size_t size, last;
 
        if (offset < 0)
                goto error;
@@ -275,6 +278,10 @@ static int cgroup_rmdir(char *name, int offset)
                goto error;
        }
 
+       /* Remove training slashes. */
+       while (last = strlen(dir) - 1, dir[last] == '/')
+               dir[last] = '\0';
+
        size = strlen(ctrl.mount_point) + strlen(dir) + 1;
        path = malloc(size);
        if (!path) {
@@ -309,9 +316,18 @@ static int cgroup_rmdir(char *name, int offset)
                }
 
                sep = strrchr(dir, '/');
-               if (!sep || sep <= dir + offset)
+               if (!sep)
                        break;
                *sep = '\0';
+
+               /* Remove training slashes and adapt separator. */
+               while (last = strlen(dir) - 1, dir[last] == '/') {
+                       sep = &dir[last];
+                       *sep = '\0';
+               }
+
+               if (sep <= dir + offset)
+                       break;
        }
 error:
        free(dir);
```